src/middlewared/middlewared/plugins/vm.py

```diff
@@ -1,487 +1,651 @@
+        ], default='1024x768'),
+        Int('vnc_port', default=None, null=True, validators=[Range(min=5900, max=65535)]),
+        Str('vnc_bind', default='0.0.0.0'),
+        Bool('wait', default=False),
+        Str('vnc_password', default=None, null=True, private=True),
+        Bool('vnc_web', default=False),
+    )
 
-        self.destroy_tap()
-        return await self.kill_bhyve_pid()
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.web_process = None
+
+    def xml(self, *args, **kwargs):
+        return create_element(
+            'graphics', type='vnc', port=str(self.data['attributes']['vnc_port']), attribute_dict={
+                'children': [
+                    create_element('listen', type='address', address=self.data['attributes']['vnc_bind']),
+                ]
+            }, **(
+                {} if not self.data['attributes']['vnc_password'] else {
+                    'passwd': self.data['attributes']['vnc_password']
+                }
+            )
+        ), create_element('controller', type='usb', model='nec-xhci'), create_element('input', type='tablet', bus='usb')
+
+    @staticmethod
+    def get_vnc_web_port(vnc_port):
+        split_port = int(str(vnc_port)[:2]) - 1
+        return int(str(split_port) + str(vnc_port)[2:])
+
+    def post_start_vm(self, *args, **kwargs):
+        vnc_port = self.data['attributes']['vnc_port']
+        vnc_bind = self.data['attributes']['vnc_bind']
+        vnc_web_port = self.get_vnc_web_port(vnc_port)
+
+        web_bind = f':{vnc_web_port}' if vnc_bind == '0.0.0.0' else f'{vnc_bind}:{vnc_web_port}'
+        self.web_process = subprocess.Popen(
+            [
+                '/usr/local/libexec/novnc/utils/websockify/run', '--web',
+                '/usr/local/libexec/novnc/', '--wrap-mode=ignore', web_bind, f'{vnc_bind}:{vnc_port}'
+            ], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL,
+        )
 
-    async def running(self):
-        bhyve_error = await (await Popen(['bhyvectl', '--vm={}'.format(str(self.vm['id']) + '_' + self.vm['name'])], stdout=subprocess.PIPE, stderr=subprocess.PIPE)).wait()
-        if bhyve_error == 0:
-            if self.proc:
-                try:
-                    os.kill(self.proc.pid, 0)
-                except OSError:
-                    self.logger.error('===> VMM {0} is running without bhyve process.'.format(self.vm['name']))
-                    return False
-                return True
-            else:
-                # XXX: We return true for now to keep the vm.status sane.
-                # It is necessary handle in a better way the bhyve process associated with the vmm.
-                return True
-        elif bhyve_error == 1:
-            return False
+    def post_stop_vm(self, *args, **kwargs):
+        if self.web_process and psutil.pid_exists(self.web_process.pid):
+            os.kill(self.web_process.pid, signal.SIGTERM)
```

Comment:
We have `service.terminate_process` that also sends SIGKILL after specific timeout.

Reply:
But we can't access middlewared from here, VMSupervisor can act as a standalone and doesn't need middlewared at all right now. Do you think that aspect needs to be changed ?

Reply:
I doubt it will ever be used as standalone but making no leftover processes is a good reason to inject middleware here.